packages/server/src/utils/gpu-setup.ts

```diff
@@ -0,0 +1,276 @@
+					path: "nvidia-container-runtime",
+					runtimeArgs: [],
+				},
+			},
+			"default-runtime": "nvidia",
+			"node-generic-resources": [`GPU=${initialStatus.availableGPUs}`],
+		};
+
+		// Different commands for local and remote setup
+		if (serverId) {
+			// Remote server setup (using sudo)
+			const setupCommands = [
+				"sudo -n true",
+				`echo '${JSON.stringify(daemonConfig, null, 2)}' | sudo tee /etc/docker/daemon.json`,
+				"sudo mkdir -p /etc/nvidia-container-runtime",
+				'echo "swarm-resource = \\"DOCKER_RESOURCE_GPU\\"" | sudo tee -a /etc/nvidia-container-runtime/config.toml',
+				"sudo systemctl daemon-reload",
+				"sudo systemctl restart docker",
+			].join(" && ");
+
+			await execAsyncRemote(serverId, setupCommands);
+		} else {
+			// Local server setup (using pkexec for GUI password prompt)
+			const configFile = `/tmp/docker-daemon-${Date.now()}.json`;
+			await fs.writeFile(configFile, JSON.stringify(daemonConfig, null, 2));
+
+			const setupCommands = [
+				// Use pkexec for GUI password prompt
+				`pkexec sh -c '
+					cp ${configFile} /etc/docker/daemon.json && 
+					mkdir -p /etc/nvidia-container-runtime && 
+					echo "swarm-resource = \\"DOCKER_RESOURCE_GPU\\"" >> /etc/nvidia-container-runtime/config.toml && 
+					systemctl daemon-reload && 
+					systemctl restart docker
+				'`,
+				`rm ${configFile}`, // Clean up temp file
+			].join(" && ");
+
+			await execAsync(setupCommands);
+		}
+
+		// 4. Reduced wait time for Docker restart
+		await new Promise((resolve) => setTimeout(resolve, 10000));
+
+		// 5. Add GPU label to the node
+		const labelCommand = `docker node update --label-add gpu=true ${nodeId.trim()}`;
+		if (serverId) {
+			await execAsyncRemote(serverId, labelCommand);
+		} else {
+			await execAsync(labelCommand);
+		}
+
+		// 6. Quick final verification
+		await new Promise((resolve) => setTimeout(resolve, 5000));
+		const finalStatus = await checkGPUStatus(serverId);
+
+		if (!finalStatus.swarmEnabled) {
+			const diagnosticCommands = [
+				`docker node inspect ${nodeId.trim()}`,
+				'nvidia-smi -a | grep "GPU UUID"',
+				"cat /etc/docker/daemon.json",
+				"cat /etc/nvidia-container-runtime/config.toml",
+			].join(" && ");
+
+			const { stdout: diagnostics } = serverId
+				? await execAsyncRemote(serverId, diagnosticCommands)
+				: await execAsync(diagnosticCommands);
+
+			console.error("Diagnostic Information:", diagnostics);
+			throw new Error("GPU support not detected in swarm after setup");
+		}
+
+		console.log("GPU setup completed successfully:", {
+			availableGPUs: initialStatus.availableGPUs,
+			driverVersion: initialStatus.driverVersion,
+			nodeId: nodeId.trim(),
+		});
+	} catch (error) {
+		console.error("GPU Setup Error:", error);
+		if (
+			error instanceof Error &&
+			error.message.includes("password is required")
+		) {
+			throw new Error(
+				"Sudo access required. Please run with appropriate permissions.",
+			);
+		}
+		throw error;
+	}
+}
```

Comment:
The same as above, it would be ideal to create certain functionalities in small portions of functions, so it becomes easier to understand